Stabilizer causes type error in combination with type member

This fails to compile on 2.13:

```scala
// class Foo[FT] {
class Foo {
  type FT
  class I
  def m(b: FT, o: Option[I]): Int = 0
}

object Test {
  // def f[T]: Foo[FT] = ???
  def f[T]: Foo { type FT = T } = ???
  def t = {
    val b: Any = ???
    f.m(b, None)
  }
}
```

The compiler inserts a stabilizer (https://github.com/scala/scala/pull/5999) because the signature of `f.m` refers to `Foo.this.I`:

```
        <synthetic> <stable> <artifact> val stabilizer$1: Foo{type FT = T} = Test.this.f[T];
        stabilizer$1.m(b, None)
```

The `T` in the stabilizer's type is a reference to the type param of `f`. The type error is

```
A.scala:11: error: type mismatch;
 found   : Any
 required: T
    f.m(b, None)
        ^
```

The variant (in comments) using a type parameter `FT` instead of a type member compiles correctly.

The difference is how the parameter type `FT` is represented in the method `stabilizer$1.m` (after asSeenFrom):
  - when `FT` is a type param, `stabilizer$1.m` has parameter type `TypeRef` to `T`, the undetermined type param of `f`. `FT` is substituted.
  - when `FT` is a type member, `m`'s parameter type is an `AliasNoArgsTypeRef` with prefix `stabilizer$1.type` and symbol `FT` (it's not substitued). It's `relativeInfo` indeed points to `T`, the undetermined type param of `f`.

Now the issue is that [`instantiateTypeParams`](https://github.com/scala/scala/blob/v2.13.13/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L3946) and [`inferMethodInstance`](https://github.com/scala/scala/blob/v2.13.13/src/compiler/scala/tools/nsc/typechecker/Typers.scala#L3965) use substSym to replace references to the type param `T`, but that leaves the `AliasNoArgsTypeRef(stabilizer$1.type, FT)` untouched.

----

I think this is specific to stabilizers, because the compiler inserts a stable `val` but leaves some type params of its rhs undetermined. That's not possible to write in source, `val stabilizer$1: Foo{type FT = T}` would need a concrete type for `T`.

So maybe there's some workaround we can do when creating the stabilizer. Changing methods like substSym / instantiateTypeParams / inferMethodInstance for AliasTypeRefs would likely break things (`normalize` gives the `relativeInfo`, but that probably drops too much information if we call it that widely).

--- a/src/compiler/scala/tools/nsc/typechecker/Typers.scala
+++ b/src/compiler/scala/tools/nsc/typechecker/Typers.scala
@@ -5524,7 +5524,12 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
               context.pendingStabilizers ::= vdef
               qual.changeOwner(context.owner -> vsym)
               val newQual = Ident(vsym) setType singleType(NoPrefix, vsym) setPos qual.pos.focus
-              return typedSelect(tree, newQual, name)
+              return typedSelect(tree, newQual, name).modifyType(_.map {
+                // very specific fix for scala/bug#12987 (details in the ticket)
+                case t: AliasTypeRef if t.pre.termSymbol == vsym && context.undetparams.contains(t.normalize.typeSymbol) =>
+                  t.normalize
+                case t => t
+              })
             }
 
             val tree1 = tree match {

--- a/test/files/pos/t12987.scala
+++ b/test/files/pos/t12987.scala
@@ -1,0 +1,30 @@
+object typeMember {
+  class Foo {
+    type FT
+    class I
+    def m(b: FT, o: Option[I]): Int = 0
+  }
+
+  object Test {
+    def f[T]: Foo { type FT = T } = ???
+    def t = {
+      val b: Any = ???
+      f.m(b, None)
+    }
+  }
+}
+
+object typeParam {
+  class Foo[FT] {
+    class I
+    def m(b: FT, o: Option[I]): Int = 0
+  }
+
+  object Test {
+    def f[T]: Foo[T] = ???
+    def t = {
+      val b: Any = ???
+      f.m(b, None)
+    }
+  }
+}

